### PR TITLE
Azure DevOps support

### PR DESCRIPTION
**Pull request by devandanger**
Azure DevOps provides all capabilities for Danger in theory.
Has source repositories.
Has PR capabilities.
Has CI driving it.

Danger currently supports mostly things around github/bitbucket/circle/travis.  Need to generacize these capabilities to make it easier to add-in Azure Devops.

- [ ] Reposlug functionality is geared only to Azure Devops support
- [ ] CI Type is only for travis / circle. Unclear where this capability is required.
- [ ] PR process only for danger.

### CODE REVIEW

**Comment by orta** [comment]
Yep, in theory you could have Azure Devops as either a CI talking to GitHub or talking to itself. The CI ostorye is easy and looks like you're mostly there already. 

Making danger talk back, and provide a rich API for PRs on devops is a pretty solid amount of work, and you might want to consider trying to do it incrementally. But I'm very open to supporting it 👍

**Comment by devandanger** [comment]
Now looking at the needs of the API I see your point. Having a feature parity for the PR APIs will not be a goal of this PR.  If I can get one example from the DSL working like this 'warning this is a giant PR' that be cool.

I need to do some more testing on the actual API approach since there are more constraints to consider.  

In general, I'm more in the explorative mode so this is really in draft form still.

Appreciate the support though!

File: source/commands/init/setup-bot-account.ts [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    ui.say("be writing comments, and you do not need special access for that.")
+  } else {
+    ui.say(`${state.botName} will need access to your repo. Simply because the code is not available for the public`)
+    ui.say("to read and comment on.")
+  }
+
+  await ui.pause(1)
+  ui.say("\nCool, please press return when you have your account ready (and you've verified the email...)")
+  ui.waitForReturn()
+}
```
Comment (orta) [inline source/commands/init/setup-bot-account.ts]: hah - cool! No one else has tried to add their own version of this!

**Comment by 9jaboy** [comment]
any update on this?

**Comment by 9jaboy** [comment]
Is this PR ready to go?

**Comment by orta** [comment]
Nope, this PR has a long way to go - given that it's been a year, it might be worth closing at this point.

**Comment by 9jaboy** [comment]
ok, thanks for your response

**Comment by pavelhynek** [comment]
any update on this?

**Comment by fbartho** [comment]
@pavelhynek @schirrel as documented in [September of 2020](https://github.com/danger/danger-js/pull/925#issuecomment-700617196), this PR wasn't ready to ship and doesn't appear to be actively in progress, and was probably worth closing.

To avoid continued confusion in the public, I'm going to close this for now, until an interested contributor wants to try to pick it back up again.

**Comment by krishnasaga** [comment]
I managed to write a function to use Azure DevOps REST API to get the diff between two branches

`javascript

export const getDiff = async ({
  repository,
  baseVersion,
  targetVersion,
  authKey,
}) => {
  const result = await fetch(
    `${AZURE_ORGANIZATION_URL}/_apis/git/repositories/${repository}/diffs/commits?baseVersion=${baseVersion}&targetVersion=${targetVersion}&api-version=7.1-preview.1`,
    {
      method: 'GET',
      headers: {
        Authorization: `Basic ${authKey} `,
      },
    }
  );

  const json = await result.json();

  return json;
};
`

**Pull request closed without merging**